compiler/rustc_hir_typeck/src/op.rs

```diff
@@ -148,10 +148,11 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         rhs_ty,
                         op,
                     );
-                    self.demand_suptype(expr.span, builtin_return_ty, return_ty);
+                    self.demand_eqtype(expr.span, builtin_return_ty, return_ty);
+                    builtin_return_ty
```

Comment:
But returning `builtin_return_ty` here means that when we equate `_#0i` and `_#1i`, the type we return will be `_#0i` instead of `<#_0i as Add<_#i1>>::Output`, which helps guide inference in the solver.